Add implementation of `conv1d_transpose`

yongtang opened the pull request:
This fix tries to address the issue raised in #8729 by providing the implementation of `conv1d_transpose`.

This fix fixes #8729.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add implementation of `conv1d_transpose`

This fix tries to address the issue raised in 8729 by
providing the implementation of `conv1d_transpose`.

This fix fixes 8729.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for `conv1d_transpose`

This fix adds test case for `conv1d_transpose`.

This fix is related to 8729.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

sb2nov commented:
Jenkins, test this please.

sb2nov commented:
@yongtang can you look at the test failure

yongtang committed:
Fix xrange issue in Python 3

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @sb2nov. The failure was caused by `xrange` in python 3. The failure has been fixed and the PR has been updated now. Please take a look and let me know if there are any other issues.

sb2nov commented:
Jenkins, test this please.

yongtang commented:
From the Jenkins output the Python 3 build has been fixed. The “Linux CPU Test Makefile” might be caused by the network issue. I think a rerun will fix it.

alextp commented:
Jenkins, test this please.

alextp commented on tensorflow/python/ops/nn_ops.py:
```diff
@@ -2043,6 +2044,98 @@ def conv1d(value, filters, stride, padding,
+  Args:
+    value: A 3-D `Tensor` of type `float` and shape
+      `[batch, in_width, in_channels]` for `NHWC` data format or
+      `[batch, in_channels, in_width]` for `NCHW` data format.
```
I'm confused; if it's conv1d, why is there h and w?

yongtang commented:
Thanks @alextp for the review and sorry for the late reply as I missed the GitHub previously notification. Indeed `NHWC` and `NCHW` is not appropriate for 1d situations. I have rebased and updated the PR so that `NWC` and `NCW` are used instead.

Please take a look and let me know if there are any issues.

alextp reviewed (changes requested):
First, why the diff in fast_tensor_util.cpp? Is that a merger artifact?

Second, let's add an import of this symbol to tf.contrib.nn ; see how it's done in nth_element: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/nn/__init__.py#L35

This way users can access this function through the contrib module for at least one release before we graduate it to tf.nn.

yongtang committed:
Change allowed data format from `NHWC`/`NCHW` -> `NWC`/`NCW`

as 1d does not have height.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Expose conv1d_transpose from tf.contrib.nn

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @alextp for the help.

Not sure why `fast_tensor_util.cpp` was included in the PR as the filename is actually in `.gitignore`. I have removed the file from PR manually.

The conv1d_transpose has also been imported in `tf.contrib.nn` and exposed. Please take a look.

alextp commented:
Jenkins, test this please.

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
Failures are unrelated.
